### ISSUE TITLE
check for offline endpoints

### ISSUE THREAD

**Issue by MarcusEbert** [hep-gc/dynafed_storagestats#24]
If there is an endpoint currently offline, the program hangs (forever? wait only some minutes).    
Needs to have a check for offline endpoints and trying to reach only online endpoints.

**Comment by ffgalindo** [comment]
For what type of endpoint did you see this happen?

**Comment by MarcusEbert** [comment]
That was for minio endpoints where the machine didn't run anymore or where the minio process didn't run anymore. (in that case for the one on  cc-east and cc-west)  Probably happens when any machine is down or nothing listening on the correct port anymore?

**Comment by ffgalindo** [comment]
I think polling memcached to use the online/offline is a good idea. I'll think how to work it into the logic.

For the timeout, rather than the latency, using what is setup as the conn_timeout option would make more sense. And if that is not set, to default it to 5 seconds or so.

________________________________

**Comment by ffgalindo** [comment]
I see, you mean the actual latency measured for the endpoint by UGR. I thought you meant the max_latency setting. OK, I can see the appeal for that. I wonder if it would work well with endpoints with really latency, <100 ms. I suppose it should, but maybe it should have a lower bound of at least 1 second or so.

The conn_timeout, at least from the Dynafed manual is the "TCP connection timeout (in seconds) to use when establishing a connection to this endpoint". So I figured if it uses this, then it would be a consistent setting across Dynafed and this script.

I'll finish implementing the status check first and then decide what to use for the timeout.

**Comment by MarcusEbert** [comment]
No, I mean the max_latency setting.  The reason for that is that there are 2 possibilities:   
1) measured latency < max_latency then endpoint is used    
2) measured latency > max_latency then endpoint is not used even if it is online

So from that it seems max_latency is a good decision to be used as timeout since everything higher than max_latency will not be used anyway.   
However, if the real measured latency is just ms or in general lower than max_latency, then it is probably not a good value to be used as timeout as you mentioned. That's why I would propose the max_latency setting since it is used in the decision if an endpoint is used or not.

**Comment by ffgalindo** [comment]
OK, for now it's just using a fixed 5 seconds time out until I set it to use an option from the config, that should be coming next.
For now, it grabs Dynafed's endpoint's connection stats from memcache (if it exists) and compares the list to the ones configured and flags them to be checked or not on the basis of them being online or offline.
I've also setup multithreading to speed up the process so that it does not need to wait for each endpoint to reply and be processed to process the rest. It still waits until all of them have been checked before output of data to memcached or stdout.
I've also added a -v --verbose option so that it print on stderr the log information according to the loglevel. Regardles of --stdout flag, but if this is set then the stats will be printed at the end of the run.
